MINOR: Insure that KafkaStreams client is closed if test fails

mjsax opened the pull request:
`KafkaStreams` instances must be closed even if we did not call `#start()`.

Some additional Java8 cleanup.

Call for review any of @guozhangwang @bbejeck @vvcephei

mjsax committed:
MINOR: Insure that KafkaStreams client is closed if test fails

satishd commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -90,24 +88,38 @@ public void before() {
+    @After
+    public void cleanup() {
+        if (streams != null) {
+            streams.close();
```
This method is invoked in happy path scenarios and it is also invoked when any method with `Before` is failed or test method throws an error. 
Is closing the stream here not sufficient instead of catching Exception explicitly in test methods and also closing them here?

satishd commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -90,24 +88,38 @@ public void before() {
 
+    @After
+    public void cleanup() {
+        if (streams != null) {
```
minor: this variable is not introduced in this PR but it may be good to rename this to `sharedStreams` or any other better name as `streams` instance is created locally in some of the tests.

satishd reviewed (commented):
@mjsax Thanks for the PR.

Nice cleanup to always close steams instance, left a comment to address a nitpick.

guozhangwang commented:
retest this please

guozhangwang commented:
Jenkins failure may be relevant:

```
[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk10-scala2.12/streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:233: 'method call lparen' has incorrect indentation level 16, expected level should be 12. [Indentation]
15:21:54 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk10-scala2.12/streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:233:17: '(' should be on the previous line. [MethodParamPad]
```

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -90,24 +88,38 @@ public void before() {
-        streams.setStateListener(stateListener);
-        Assert.assertEquals(streams.state(), KafkaStreams.State.CREATED);
-        Assert.assertEquals(stateListener.numChanges, 0);
+        try {
```
It seems we can just reuse the shared `streams` object here instead of creating a new one?

bbejeck replied:
If we reuse shared streams instance do we still need try-catch here as the `cleanup()` should close streams.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -284,14 +295,19 @@ public void testInitializesAndDestroysMetricsReporters() {
+            final int initDiff = newInitCount - oldInitCount;
+            assertTrue("some reporters should be initialized by calling on construction", initDiff > 0);
+
+            streams.start();
```
Ditto here, for this test we can still reuse the shared streams.

mjsax replied:
Not here. We need to get `oldInitCount = MockMetricsReporter.INIT_COUNT.get();` before we create `new KafkaStreams()` instance.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -491,13 +500,17 @@ public void testCannotCleanupWhileRunning() throws InterruptedException {
-                return streams.state() == KafkaStreams.State.RUNNING;
-            }
-        }, 10 * 1000, "Streams never started.");
+        try {
```
We can reuse `streams`.

bbejeck replied:
same as above with try/catch

vvcephei reviewed (commented):
Overall, LGTM.
I generally agree with @guozhangwang 's assessment that any time we do:
```
final StreamsBuilder builder = new StreamsBuilder();
final KafkaStreams streams = new KafkaStreams(builder.build(), props);
```
we might as well just use the test harness (`this.streams` instead).

bbejeck reviewed (commented):
just a couple of minor comments otherwise looks good.

bbejeck commented:
failures are checkstyle related

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -472,17 +479,19 @@ public void shouldReturnThreadMetadata() {
     @Test
     public void testCleanup() {
         final StreamsBuilder builder = new StreamsBuilder();
         final KafkaStreams streams = new KafkaStreams(builder.build(), props);
```
@guozhangwang @bbejeck Can reuse global one here, too.

mjsax committed:
Github comments
 - removed more code to increase reusage
 - fixed nameing bug
 - fixded bug that breaks incremantal building

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -149,7 +149,7 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
         materialized.withLoggingDisabled();
         final StoreBuilder storeBuilder = new KeyValueStoreMaterializer<>(materialized).materialize();
-        final String sourceName = newProcessorName(KStreamImpl.SOURCE_NAME);
+        final String sourceName = newProcessorName(KTableImpl.SOURCE_NAME);
```
@guozhangwang @bbejeck @vvcephei Discovered this by chance -- not sure if fixing this would result in compatibility issues? Thoughts?

guozhangwang replied:
Good catch.

If we are considering topic / store names only, then no. I'd suggest we just fix it right away.

vvcephei replied:
Hmm. I'm tracing through the code, but it's certainly safer just to mark it with a `//TODO` and leave it alone.

vvcephei replied:
Yeah, I think the only concern would be if the source node name were used for the store, and I don't think it is.

guozhangwang replied:
Okay, let's change it here then.

mjsax replied:
Thanks for verifying that this is a safe change/fix.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -549,15 +555,29 @@ public final void addGlobalStore(final StoreBuilder<KeyValueStore> storeBuilder,
                                      storeBuilder.loggingEnabled());
         validateTopicNotAlreadyRegistered(topic);
 
-        addGlobalStore(sourceName,
```
inlined this method

mjsax commented:
Update this. Also discovered two more bugs. See comment on `InternalStreamsBuilder` and changes on `InternalTopologyBuilder` plus new test in `InternalTopologyBuilderTest`.

enothereska commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -70,12 +68,12 @@
-    // We need this to avoid the KafkaConsumer hanging on poll (this may occur if the test doesn't complete
-    // quick enough)
+    // We need this to avoid the KafkaConsumer hanging on poll
+    // (this may occur if the test doesn't complete quick enough)
```
nit: quick->quickly

guozhangwang reviewed (approved):
LGTM. Please feel free to merge.

vvcephei commented on streams/src/test/java/org/apache/kafka/test/MockStateStore.java:
```diff
@@ -21,10 +21,13 @@
+import java.util.List;
 
-public class MockStateStore implements StateStore {
+public class MockStateStore implements KeyValueStore {
```
`MockStateStore` -> `MockKeyValueStore` ?

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -87,59 +85,64 @@ public void before() {
+        TestUtils.waitForCondition(
+            () -> streams.state() == KafkaStreams.State.NOT_RUNNING,
+            10 * 1000,
+            "Streams never stopped.");
```
Just out of curiosity, how does this differ from the following?
```
assertTrue(streams.close(10, TimeUnit.SECONDS));
```

mjsax replied:
Not sure... Seems we could rewrite it. \cc @guozhangwang @bbejeck WDYT?

bbejeck replied:
none from what I can see, but I'm not sure it's worth holding up the PR for it.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -269,111 +263,112 @@ public void testLocalThreadCloseWithoutConnectingToBroker() {
+        try {
+            streams.start();
+        } finally {
+            streams.close();
```
The fact that we are having to do this in our tests suggests to me that maybe we should think about making KafkaStreams `AutoCloseable`. Should I go ahead and make a jira for this?

mjsax replied:
In general I am in favor to this. \cc @guozhangwang @bbejeck Thoughts?

bbejeck replied:
+1 for me on `AutoCloseable`

vvcephei commented on streams/src/test/java/org/apache/kafka/test/MockStoreBuilder.java:
```diff
@@ -18,10 +18,10 @@
 import org.apache.kafka.streams.state.internals.AbstractStoreBuilder;
 
-public class MockStoreBuilder extends AbstractStoreBuilder<Integer, byte[], StateStore> {
+public class MockStoreBuilder extends AbstractStoreBuilder<Integer, byte[], KeyValueStore> {
```
Also here, should it be `MockKeyValueStoreBuilder`?

mjsax committed:
Github comments

mjsax commented:
Updated this.

vvcephei reviewed (approved):
LGTM. Thanks, @mjsax !
